Sources/SwiftJWT/JWTEncoder.swift

```diff
@@ -0,0 +1,219 @@
+ do {
+    let jwtString = try rsaJWTEncoder.encodeToString(jwt)
+ } catch {
+    print("Failed to encode JWT: \(error)")
+ }
+ ```
+ */
+public class JWTEncoder: BodyEncoder {
+    
+    let keyIDToSigner: (String) -> JWTSigner?
+    let jwtSigner: JWTSigner?
+    let header: Header?
+    
+    // MARK: Initializers
+    
+    /// Initialize a `JWTEncoder` instance with a single `JWTSigner`.
+    ///
+    /// - Parameter algorithm: The `Algorithm` that will be used to sign the JWT.
+    /// - Returns: A new instance of `JWTEncoder`.
+    public init(jwtSigner: JWTSigner, header: Header? = nil) {
+        self.keyIDToSigner = {_ in return jwtSigner }
+        self.jwtSigner = jwtSigner
+        self.header = header
+    }
+    
+    /// Initialize a `JWTEncoder` instance with a function to generate the `JWTSigner` from the JWT `kid` header.
+    ///
+    /// - Parameter algorithm: The `Algorithm` that will be used to sign the JWT.
+    /// - Returns: A new instance of `JWTEncoder`.
+    public init(keyIDToSigner: @escaping (String) -> JWTSigner?, header: Header? = nil) {
+        self.keyIDToSigner = keyIDToSigner
+        self.header = header
+        self.jwtSigner = nil
+    }
+    
+    // MARK: Encode
+    
+    /// Encode a `JWT` instance into a utf8 encoded JWT String.
+    ///
+    /// - Parameter value: The JWT instance to be encoded as Data.
+    /// - Returns: The utf8 encoded JWT String.
+    /// - throws: `JWTError.invalidUTF8Data` if the provided Data can't be decoded to a String.
+    /// - throws: `JWTError.invalidKeyID` if the KeyID `kid` header fails to generate a jwtSigner.
+    /// - throws: `EncodingError` if the encoder fails to encode the object as Data.
+    public func encode<T : Encodable>(_ value: T) throws -> Data {
+        guard let jwt = try self.encodeToString(value).data(using: .utf8) else {
+            throw JWTError.invalidUTF8Data
+        }
+        return jwt
+    }
+    
+    /// Encode a `JWT` instance as a JWT String.
+    ///
+    /// - Parameter value: The JWT instance to be encoded as a JWT String.
+    /// - Returns: A JWT String.
+    /// - throws: `JWTError.invalidKeyID` if the KeyID `kid` header fails to generate a jwtSigner.
+    /// - throws: `EncodingError` if the encoder fails to encode the object as Data.
+    public func encodeToString<T : Encodable>(_ value: T) throws -> String {
+        let encoder = _JWTEncoder()
+        try value.encode(to: encoder)
+        var _header: Header
+        if let header = header {
+            _header = header
+        } else {
+            guard let headerString = encoder.header, let headerData = Data(base64urlEncoded: headerString) else {
+                throw EncodingError.invalidValue(value, EncodingError.Context(codingPath: [], debugDescription: "Failed to encode into header CodingKey"))
+            }
+            let newHeader = try JSONDecoder().decode(Header.self, from: headerData)
```

Comment:
Similar to the argument for Decoder, this code ends up decoding the Header string that we just encoded, in order to fish out the `kid` header.  Could we move this code into the _JWTEncoder to avoid doing this extra work?